### PR TITLE
`[lib]` Add a library for accessing parallelcluster

### PR DESCRIPTION
**Pull request by charlesg3**
### Description of changes
* Describe *what* you're changing and *why* you're doing these changes.

This commit adds a library for accessing parallelcluster by generating the class from the spec. It does more type checking as the CLI is able to do that portion through the argument parsing library, as well as accepting Python datastructures for those parameters of type "file" such as cluster and image configuration.

### Tests

Manual tests for exercising the code against AWS resources:

```.py
#!/usr/bin/env python3
# Copyright 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License"). You may not
# use this file except in compliance with the License. A copy of the License is
# located at
#
# http://aws.amazon.com/apache2.0/
#
# or in the "LICENSE.txt" file accompanying this file. This file is distributed
# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, express or
# implied. See the License for the specific language governing permissions and
# limitations under the License.

import pprint
import time

import yaml
from pcluster.lib import ParallelCluster

CONFIG_TEMPLATE = "cluster-config.yaml.templ"
HEAD_NODE_SUBNET = "[REDACTED]"
COMPUTE_NODE_SUBNET = "[REDACTED]"
KEYNAME = "[REDACTED]"
CLUSTER_NAME = "mycluster"

pp = pprint.PrettyPrinter(indent=2)

def cluster_config():
    with open(CONFIG_TEMPLATE, "r", encoding="utf-8") as inf:
        config = yaml.loads(inf.read())

    config["HeadNode"]["Networking"]["SubnetId"] = COMPUTE_NODE_SUBNET
    config["HeadNode"]["Ssh"]["KeyName"] = KEYNAME
    queue0 = config["Scheduling"]["SlurmQueues"][0]
    queue0["ComputeResources"][0]["Networkin"]["SubnetIds"][0] = COMPUTE_NODE_SUBNET
    return config

def cluster_log_test(pc):
    log_streams = pc.list_cluster_log_streams(cluster_name=CLUSTER_NAME)
    assert len(log_streams['logStreams']) > 0

    stream_name = log_streams['logStreams'][0]['logStreamName']

    log_events = pc.get_cluster_log_events(cluster_name=CLUSTER_NAME, log_stream_name=stream_name)

    assert len(log_events["events"]) > 0

def cluster_fleet_test(pc):
    pc.update_compute_fleet(cluster_name=CLUSTER_NAME,status="STOP_REQUESTED")

    fleet_status = pc.describe_compute_fleet(cluster_name=CLUSTER_NAME)["status"]
    while fleet_status == "STOP_REQUESTED":
        print("Waiting to stop compute fleet...")
        time.sleep(1)
        fleet_status = pc.describe_compute_fleet(cluster_name=CLUSTER_NAME)["status"]

    pc.update_compute_fleet(cluster_name=CLUSTER_NAME,status="START_REQUESTED")

    fleet_status = pc.describe_compute_fleet(cluster_name=CLUSTER_NAME)["status"]
    while fleet_status == "START_REQUESTED":
        print("Waiting to start compute fleet...")
        time.sleep(1)
        fleet_status = pc.describe_compute_fleet(cluster_name=CLUSTER_NAME)["status"]

    assert pc.describe_compute_fleet(cluster_name=CLUSTER_NAME)["status"] in {"STARTING", "RUNNING"}

def cluster_test(config):
    pc = ParallelCluster()

    print(f"Creating cluster [{CLUSTER_NAME}]...")
    pp.pprint(pc.create_cluster(cluster_name=CLUSTER_NAME, cluster_configuration=config, wait=True))

    status = pc.describe_cluster(cluster_name=CLUSTER_NAME)
    while status['clusterStatus'] == "CREATE_IN_PROGRESS":
        time.sleep(1)
        status = pc.describe_cluster(cluster_name=CLUSTER_NAME)

    cluster = pc.list_clusters(query=f"clusters[?clusterName=='{CLUSTER_NAME}']|[0]")
    assert cluster['clusterName'] == CLUSTER_NAME

    cluster_log_test(pc)
    cluster_fleet_test(pc)

    assert len(pc.describe_cluster_instances(cluster_name=CLUSTER_NAME)["instances"]) > 0
    assert len(pc.get_cluster_stack_events(cluster_name=CLUSTER_NAME)["events"]) > 0
    pp.pprint((pc.delete_cluster(cluster_name=CLUSTER_NAME)))

def image_test():
    pc = ParallelCluster()
    assert len(pc.list_images(image_status="AVAILABLE")["images"]) > 0
    assert pc.describe_image(image_id=images[0]["imageId"])["imageBuildStatus"] == "BUILD_COMPLETE"

if __name__ == "__main__":
    cluster_test(cluster_config())
    image_test()
```

`cluster-config.yaml.templ`:

```.yaml
Image:
  Os: alinux2
HeadNode:
  InstanceType: t2.large
  Networking:
    SubnetId: ${HEAD_NODE_SUBNET}
  Ssh:
    KeyName: ${KEYNAME}
Scheduling:
  Scheduler: slurm
  SlurmQueues:
  - Name: queue0
    ComputeResources:
    - Name: queue0-i0
      InstanceType: t2.micro
      MinCount: 0
      MaxCount: 10
    Networking:
      SubnetIds:
      - ${COMPUTE_NODE_SUBNET}
```

* Describe the added/modified tests.
Unit tests are added to test the parameter validation and resolution to the underlying code and type coersion.

### Checklist
- [x] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [x] Check all commits' messages are clear, describing what and why vs how.
- [x] Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- [x] Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: cli/src/pcluster/lib/__init__.py [hunk 1]
```diff
@@ -0,0 +1,10 @@
+# Copyright 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
+# with the License. A copy of the License is located at http://aws.amazon.com/apache2.0/
+# or in the "LICENSE.txt" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES
+# OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions and
+# limitations under the License.
+
+# flake8: noqa
```
Comment (demartinofra) [inline cli/src/pcluster/lib/__init__.py:10]: why is the suppression needed? please leave a comment if we can't enable it

File: cli/src/pcluster/lib/lib.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+import yaml
+
+import pcluster.api.controllers.cluster_operations_controller
+import pcluster.api.errors
+import pcluster.cli.model
+from pcluster.cli.entrypoint import dispatch, param_coerce
+from pcluster.utils import to_snake_case
+
+
+def _gen_class(model):
```
Comment (demartinofra) [inline cli/src/pcluster/lib/lib.py]: [optional] adding typing info to this module would improve navigating through the generation functions.

**Review by demartinofra — DISMISSED** [review]
LGTM. Left 2 non-blocking comments.

File: cli/src/pcluster/lib/lib.py (hunk 2 above)
Comment (charlesg3) [reply cli/src/pcluster/lib/lib.py]: I added some typing -- the real help is explained in the other file that generates the model so I added a comment pointing to that (as well).

File: cli/src/pcluster/lib/__init__.py (hunk 1 above)
Comment (charlesg3) [reply cli/src/pcluster/lib/__init__.py:10]: This is so we can include `ParallelCluster` class even though it isn't used in the `__init__.py` file -- I added a comment explaining that.

**Review by demartinofra — APPROVED** [review]

**Pull request merged**